File: android/src/main/java/com/smileidentity/react/expo/SmileIDSmartSelfieAuthenticationView.kt
package com.smileidentity.react.expo

import android.content.Context
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.systemBarsPadding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.Modifier
import com.smileidentity.SmileID
import com.smileidentity.compose.SmartSelfieAuthentication
import com.smileidentity.results.SmartSelfieResult
import com.smileidentity.results.SmileIDResult
import com.smileidentity.util.randomJobId
import com.smileidentity.util.randomUserId
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher

/**
 * Smart Selfie Authentication View using ExpoView
 **/
class SmileIDSmartSelfieAuthenticationView(context: Context, appContext: AppContext) :
    SmileIDExpoComposeView(
        context = context,
        appContext = appContext,
        shouldHostComposeContent = true
    ) {
    private var props = mutableStateOf(SmartSelfieProps())
    private val onResult by EventDispatcher()
    private val onError by EventDispatcher()

    @Composable
    override fun Content() {
        SmartSelfieAuthenticationView(
            props = props.value,
            onResult = { result ->
                onResult(
                    mapOf(
                        "selfieFile" to result.selfieFile.toString(),
                        "livenessFiles" to result.livenessFiles.toString()
                    )
                )
            },
            onError = { error ->
                onError(
                    mapOf(
                        "error" to error.localizedMessage.toString()
                    )
                )
            }
        )
    }

    fun updateConfig(config: SmartSelfieParams) {
        props.value = config.toSmartSelfieProps()
    }
}

/**
 * Compose view that wraps the SmileID SmartSelfie authentication screen
 **/
@Composable
private fun SmartSelfieAuthenticationView(
    props: SmartSelfieProps,
    onResult: (SmartSelfieResult) -> Unit,
    onError: (Throwable) -> Unit
) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .navigationBarsPadding()
            .systemBarsPadding()
    ) {
        SmileID.SmartSelfieAuthentication(
            userId = props.userId ?: randomUserId(),
            jobId = props.jobId ?: randomJobId(),
            allowNewEnroll = props.allowNewEnroll,
            allowAgentMode = props.allowAgentMode,
            showAttribution = props.showAttribution,
            showInstructions = props.showInstructions,
            skipApiSubmission = props.skipApiSubmission,
            smileSensitivity = props.smileSensitivity,
            extraPartnerParams = props.extraPartnerParams,
        ) { result ->
            when (result) {
                is SmileIDResult.Success -> {
                    onResult(result.data)
                }

                is SmileIDResult.Error -> {
                    onError(result.throwable)
                }
            }
        }
    }
}

File: ios/SmileIDSmartSelfieAuthenticationView.swift
import ExpoModulesCore
import SwiftUI
import SmileID

// SmartSelfie Authentication View using ExpoView
final class SmileIDSmartSelfieAuthenticationView: ExpoView {
    let onResult = EventDispatcher()
    let onError = EventDispatcher()
    private let delegate: SmartSelfieDelegate
    private let hostingController: UIHostingController<SmartSelfieAuthenticationView>
    private var config: SmartSelfieParams?

    required init(appContext: AppContext? = nil) {
        delegate = SmartSelfieDelegate()
        hostingController = UIHostingController(
            rootView: SmartSelfieAuthenticationView(
                delegate: delegate,
                config: nil
            )
        )
        // Force light mode
        hostingController.overrideUserInterfaceStyle = .light

        super.init(appContext: appContext)

        // Set up delegate callbacks
        delegate.onResult = { [weak self] result in
            self?.onResult(result)
        }

        delegate.onError = { [weak self] error in
            self?.onError(["error": error.localizedDescription])
        }

        // Add the hosting controller's view
        addSubview(hostingController.view)
        hostingController.view.fillSuperview()
    }

    func updateConfig(_ config: SmartSelfieParams) {
        self.config = config
        hostingController.rootView = SmartSelfieAuthenticationView(
            delegate: delegate,
            config: config
        )
    }
}

// SwiftUI view that wraps the SmileID SmartSelfie authentication view
struct SmartSelfieAuthenticationView: View {
    let delegate: SmartSelfieDelegate
    let config: SmartSelfieParams?

    var body: some View {
        if let config = config {
            SmileID.smartSelfieAuthenticationScreen(
                userId: config.userId ?? generateUserId(),
                jobId: config.jobId ?? generateJobId(),
                allowNewEnroll: config.allowNewEnroll,
                allowAgentMode: config.allowAgentMode,
                forceAgentMode: config.forceAgentMode,
                showAttribution: config.showAttribution,
                showInstructions: config.showInstructions,
                useStrictMode: config.useStrictMode,
                smileSensitivity: SmileSensitivity(from: config.smileSensitivity),
                skipApiSubmission: config.skipApiSubmission,
                extraPartnerParams: config.extraPartnerParams,
                delegate: delegate
            )
        } else {
            Text("Configuration not provided").foregroundColor(.red)
        }
    }
}
